dockerd-entrypoint.sh

```diff
@@ -148,10 +148,14 @@ if [ "$1" = 'dockerd' ]; then
 		# https://github.com/docker-library/docker/issues/350
 		# https://github.com/moby/moby/issues/26824
 		# https://github.com/docker-library/docker/pull/437#issuecomment-1854900620
-		if ! modprobe nf_tables; then
+		modprobe nf_tables || :
+		if ! iptables -nL > /dev/null 2>&1; then
+			# might be host has no nf_tables, but Alpine is all-in now (so let's try a legacy fallback)
 			modprobe ip_tables || :
-			# see https://github.com/docker-library/docker/issues/463 (and the dind Dockerfile where this directory is set up)
-			export PATH="/usr/local/sbin/.iptables-legacy:$PATH"
+			if /usr/local/sbin/.iptables-legacy/iptables -nL > /dev/null 2>&1; then
+				# see https://github.com/docker-library/docker/issues/463 (and the dind Dockerfile where this directory is set up)
+				export PATH="/usr/local/sbin/.iptables-legacy:$PATH"
+			fi
 		fi
```

Comment:
When host has both `nf_tables` and `ip_tables` but is using `ip_tables`, `iptables` is giving this warning:

```
iptables -nL >/dev/null
# Warning: iptables-legacy tables present, use iptables-legacy to see them
```
Something like this worked for me to detect that case and use `iptables-legacy`:

```
        elif iptables -nL 2>&1 >/dev/null | grep -q tables-legacy; then
                export PATH="/usr/local/sbin/.iptables-legacy:$PATH"
```

Reply:
I found the code in `iptables` itself responsible for that check: :smile:

https://git.netfilter.org/iptables/tree/iptables/nft-shared.c?id=f5cf76626d95d2c491a80288bccc160c53b44e88#n420

It's effectively "read from `/proc/net/ip_tables_names` and if it's non-empty, print the warning", so we can do better than grepping the output of the tool and can instead do a simple `[ -s /proc/net/ip_tables_names ]` (although we probably ought to do all three of those "legacy" files the upstream code references, for good measure).

Reply:
> Oh, interesting! Any idea in what situations/use cases a fresh network namespace might have legacy tables set up in it already?

Did some more testing and figured this out -- it turns out that on CentOS 7, `/proc/net/ip_tables_names` is _not_ empty in a brand new network namespace, even though there aren't any actual rules.  What's even funnier is that I was confused by that upstream code reading from the files and then throwing away everything it reads, but it turns out that's done because these special kernel files have contents, but if you `stat` them their size is 0 bytes. :upside_down_face:

```console
8637e230c491:/# test -s /proc/net/ip_tables_names; echo $?
1
8637e230c491:/# stat /proc/net/ip_tables_names
  File: /proc/net/ip_tables_names
  Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
Device: 29h/41d	Inode: 4026532469  Links: 1
Access: (0440/-r--r-----)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-01-04 18:43:59.112212543 +0000
Modify: 2024-01-04 18:43:59.112212543 +0000
Change: 2024-01-04 18:43:59.112212543 +0000
8637e230c491:/# cat /proc/net/ip_tables_names
nat
mangle
security
raw
filter
```

(so my `-s` test isn't sufficient :disappointed:)